src/lib/support/tests/TestZclString.cpp

```diff
@@ -0,0 +1,160 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements a unit test suite for ZclString
+ *
+ */
+
+#include <inttypes.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <nlunit-test.h>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/UnitTestRegistration.h>
+
+#include <lib/support/ZclString.h>
+
+using namespace chip;
+using namespace chip::Logging;
+using namespace chip::Platform;
+
+static void TestZclStringLessThanMaximumSize(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t bufferMemory[256];
+    // Init zcl string
+    MutableByteSpan zclString(bufferMemory);
+    // Init c string
+    char * cString64 = static_cast<char *>(MemoryCalloc(1024, true));
+    // Add random chars to c string
+    for (size_t i = 0; i < 64; i++)
+        cString64[i] = 'A';
+
+    CHIP_ERROR err = MakeZclCharString(zclString, cString64);
+
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString64 != nullptr);
+    // ZCL String length assert
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] == 64);
+    // ZCL String data is not null
+    NL_TEST_ASSERT(inSuite, zclString.data() != NULL);
+    // ZCL String data does not hold empty string
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] != '\0');
+    // ZCL String is created with no errors
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    chip::Platform::MemoryFree(cString64);
+}
+
+static void TestZclStringEqualsMaximumSize(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t bufferMemory[256];
+    // Init zcl string
+    MutableByteSpan zclString(bufferMemory);
+    // Init c string
+    char * cString254 = static_cast<char *>(MemoryCalloc(1024, true));
+    // Add random chars to c string
+    for (size_t i = 0; i < 254; i++)
+        cString254[i] = 'A';
+
+    CHIP_ERROR err = MakeZclCharString(zclString, cString254);
+
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString254 != nullptr);
+    // ZCL String length assert
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] == 254);
+    // ZCL String data is not null
+    NL_TEST_ASSERT(inSuite, zclString.data() != NULL);
+    // ZCL String data does not hold empty string
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] != '\0');
+    // ZCL String is created with no errors
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    chip::Platform::MemoryFree(cString254);
+}
+
+static void TestZclStringBiggerThanMaximumSize(nlTestSuite * inSuite, void * inContext)
```

Comment:
Should test sizes 255, 256, 257, just to make sure things are reasonable...  And ideally do something where we ensure we don't write off the end of the buffer, by actually having the buffer be a bit bigger, initializing it, claiming a smaller size to the callee, and then asserting that only the parts we expect to get written got written.

In particular, should test that if passed-in buffer is 0-sized we don't write things to it.